File: ios/ZillitLCW/Core/Network/BoxScheduleAPI.swift
import Foundation

/// BoxScheduleAPI — all Box Schedule endpoints.
/// Uses APIClient for HTTP calls with auto-injected moduledata header.
/// All write endpoints identify the user via moduledata (userId). Display name
/// is resolved client-side from the local users cache.
class BoxScheduleAPI {
    static let shared = BoxScheduleAPI()

    private let client = APIClient.shared

    // ═══════════════════════ SCHEDULE TYPES ═══════════════════════

    func getTypes() async throws -> [ScheduleType] {
        let response: APIResponse<[ScheduleType]> = try await client.request(path: "/types")
        return response.data ?? []
    }

    func createType(title: String, color: String) async throws -> ScheduleType? {
        let response: APIResponse<ScheduleType> = try await client.request(
            path: "/types", method: "POST",
            body: ["title": title, "color": color]
        )
        return response.data
    }

    func updateType(id: String, title: String? = nil, color: String? = nil) async throws -> ScheduleType? {
        var body: [String: Any] = [:]
        if let t = title { body["title"] = t }
        if let c = color { body["color"] = c }
        let response: APIResponse<ScheduleType> = try await client.request(
            path: "/types/\(id)", method: "PUT", body: body
        )
        return response.data
    }

    func deleteType(id: String) async throws {
        let _: APIResponse<EmptyData> = try await client.request(
            path: "/types/\(id)", method: "DELETE"
        )
    }

    // ═══════════════════════ SCHEDULE DAYS ═══════════════════════

    func getDays(startDate: Int64? = nil, endDate: Int64? = nil, typeId: String? = nil) async throws -> [ScheduleDay] {
        var params: [String: String] = [:]
        if let s = startDate { params["startDate"] = "\(s)" }
        if let e = endDate { params["endDate"] = "\(e)" }
        if let t = typeId { params["typeId"] = t }
        let response: APIResponse<[ScheduleDay]> = try await client.request(
            path: "/days", queryParams: params
        )
        return response.data ?? []
    }

    func createDay(
        title: String, typeId: String, dateRangeType: String,
        calendarDays: [Int64], timezone: String = "UTC",
        conflictAction: String = ""
    ) async throws -> ScheduleDay? {
        let body: [String: Any] = [
            "title": title, "typeId": typeId, "dateRangeType": dateRangeType,
            "calendarDays": calendarDays,
            "startDate": calendarDays.min() ?? 0,
            "endDate": calendarDays.max() ?? 0,
            "numberOfDays": calendarDays.count,
            "timezone": timezone,
            "conflictAction": conflictAction
        ]
        let response: APIResponse<ScheduleDay> = try await client.request(
            path: "/days", method: "POST", body: body
        )
        return response.data
    }

    func updateDay(id: String, data: [String: Any]) async throws -> ScheduleDay? {
        let response: APIResponse<ScheduleDay> = try await client.request(
            path: "/days/\(id)", method: "PUT", body: data
        )
        return response.data
    }

    /// Atomic single-day type change — PUT /days/:id/single-date.
    func updateSingleDay(id: String, date: Int64, typeId: String, action: String) async throws {
        let _: APIResponse<EmptyData> = try await client.request(
            path: "/days/\(id)/single-date", method: "PUT",
            body: ["date": date, "typeId": typeId, "action": action]
        )
    }

    func deleteDay(id: String) async throws {
        let _: APIResponse<EmptyData> = try await client.request(
            path: "/days/\(id)", method: "DELETE"
        )
    }

    func bulkUpdateDays(updates: [[String: Any]]) async throws -> [ScheduleDay] {
        let response: APIResponse<[ScheduleDay]> = try await client.request(
            path: "/days/bulk", method: "POST",
            body: ["updates": updates]
        )
        return response.data ?? []
    }

    func removeDates(entries: [[String: Any]]) async throws {
        let _: APIResponse<EmptyData> = try await client.request(
            path: "/days/remove-dates", method: "POST",
            body: ["entries": entries]
        )
    }

    func duplicateDay(sourceDayId: String, newStartDate: Int64) async throws -> ScheduleDay? {
        let response: APIResponse<ScheduleDay> = try await client.request(
            path: "/days/duplicate", method: "POST",
            body: ["sourceDayId": sourceDayId, "newStartDate": newStartDate]
        )
        return response.data
    }

    // ═══════════════════════ EVENTS ═══════════════════════

    func getEvents(
        startDate: Int64? = nil, endDate: Int64? = nil,
        scheduleDayId: String? = nil, eventType: String? = nil
    ) async throws -> [ScheduleEvent] {
        var params: [String: String] = [:]
        if let s = startDate { params["startDate"] = "\(s)" }
        if let e = endDate { params["endDate"] = "\(e)" }
        if let d = scheduleDayId { params["scheduleDayId"] = d }
        if let t = eventType { params["eventType"] = t }
        let response: APIResponse<[ScheduleEvent]> = try await client.request(
            path: "/events", queryParams: params
        )
        return response.data ?? []
    }

    func createEvent(data: [String: Any]) async throws -> ScheduleEvent? {
        let response: APIResponse<ScheduleEvent> = try await client.request(
            path: "/events", method: "POST", body: data
        )
        return response.data
    }

    func updateEvent(id: String, data: [String: Any]) async throws -> ScheduleEvent? {
        let response: APIResponse<ScheduleEvent> = try await client.request(
            path: "/events/\(id)", method: "PUT", body: data
        )
        return response.data
    }

    func deleteEvent(id: String) async throws {
        let _: APIResponse<EmptyData> = try await client.request(
            path: "/events/\(id)", method: "DELETE"
        )
    }

    // ═══════════════════════ CALENDAR ═══════════════════════

    func getCalendar(startDate: Int64? = nil, endDate: Int64? = nil) async throws -> [ScheduleDay] {
        var params: [String: String] = [:]
        if let s = startDate { params["startDate"] = "\(s)" }
        if let e = endDate { params["endDate"] = "\(e)" }
        let response: APIResponse<[ScheduleDay]> = try await client.request(
            path: "/calendar", queryParams: params
        )
        return response.data ?? []
    }

    // ═══════════════════════ ACTIVITY LOG ═══════════════════════

    struct ActivityLogData: Codable {
        let logs: [ActivityLog]
        let total: Int
        let page: Int
        let limit: Int
    }

    func getActivityLog(limit: Int = 50, page: Int = 0) async throws -> ActivityLogData {
        let response: APIResponse<ActivityLogData> = try await client.request(
            path: "/activity-log",
            queryParams: ["limit": "\(limit)", "page": "\(page)"]
        )
        return response.data ?? ActivityLogData(logs: [], total: 0, page: 0, limit: limit)
    }

    // ═══════════════════════ REVISIONS ═══════════════════════

    func getRevisions() async throws -> [Revision] {
        let response: APIResponse<[Revision]> = try await client.request(path: "/revisions")
        return response.data ?? []
    }

    func getCurrentRevision() async throws -> Revision? {
        let response: APIResponse<Revision> = try await client.request(path: "/revisions/current")
        return response.data
    }

    // ═══════════════════════ SHARE ═══════════════════════

    struct ShareLinkData: Codable {
        let token: String
        let shareUrl: String
        let attachment: Attachment?
    }

    /// POST /share/generate-link — backend renders a PDF, uploads to S3,
    /// and returns the share token + an Attachment with presigned URLs.
    ///
    /// All body fields are optional. When `watermark` is provided, both `text`
    /// and `image` may be set; the server overlays them on every page.
    func generateShareLink(
        dayIds: [String]? = nil,
        from: Int64? = nil,
        to: Int64? = nil,
        orientation: String = "landscape",
        title: String? = nil,
        watermark: [String: Any]? = nil
    ) async throws -> ShareLinkData? {
        var body: [String: Any] = ["orientation": orientation]
        if let ids = dayIds { body["dayIds"] = ids }
        if let f = from { body["from"] = f }
        if let t = to { body["to"] = t }
        if let title = title { body["title"] = title }
        if let wm = watermark { body["watermark"] = wm }

        let response: APIResponse<ShareLinkData> = try await client.request(
            path: "/share/generate-link", method: "POST",
            body: body
        )
        return response.data
    }

    struct SharedScheduleData: Codable {
        let projectId: String
        let days: [ScheduleDay]
        let events: [ScheduleEvent]
        let types: [ScheduleType]
    }

    func getSharedSchedule(token: String) async throws -> SharedScheduleData? {
        let response: APIResponse<SharedScheduleData> = try await client.request(
            path: "/share/\(token)"
        )
        return response.data
    }
}

// MARK: - Empty response helper
struct EmptyData: Codable {}

File: android/app/src/main/java/com/zillit/lcw/data/api/BoxScheduleApi.kt
package com.zillit.lcw.data.api

import android.content.Context
import com.zillit.lcw.data.model.*
import io.ktor.client.*
import io.ktor.client.call.*
import io.ktor.client.request.*
import io.ktor.client.statement.*
import io.ktor.http.*
import kotlinx.serialization.Serializable
import kotlinx.serialization.json.*

/** Thrown when /days returns 409 — UI uses this to decide whether to show the conflict dialog. */
class ScheduleConflictException(
    val conflicts: List<JsonObject>
) : RuntimeException("schedule_day_conflict")

/**
 * BoxScheduleApi — all 21 API endpoints.
 * Uses Ktor HttpClient with JsonBuilder for serialization.
 * All requests auto-include the `moduledata` auth header via KtorClient.
 */
class BoxScheduleApi(private val context: Context) {

    private val client: HttpClient by lazy { KtorClient.create(context) }
    private val baseUrl = KtorClient.BASE_URL

    // ═══════════════════════ SCHEDULE TYPES ═══════════════════════

    /** GET /types — list all schedule types (seeds defaults on first call) */
    suspend fun getTypes(): ApiResponse<List<ScheduleType>> {
        return client.get("$baseUrl/types").body()
    }

    /** POST /types — create a custom type */
    suspend fun createType(title: String, color: String): ApiResponse<ScheduleType> {
        return client.post("$baseUrl/types") {
            setBody(buildJsonObject {
                put("title", title)
                put("color", color)
            })
        }.body()
    }

    /** PUT /types/:id — update type name/color */
    suspend fun updateType(id: String, title: String? = null, color: String? = null): ApiResponse<ScheduleType> {
        return client.put("$baseUrl/types/$id") {
            setBody(buildJsonObject {
                title?.let { put("title", it) }
                color?.let { put("color", it) }
            })
        }.body()
    }

    /** DELETE /types/:id — no body; backend reads userId from moduledata. */
    suspend fun deleteType(id: String): ApiResponse<JsonObject> {
        return client.delete("$baseUrl/types/$id").body()
    }

    // ═══════════════════════ SCHEDULE DAYS ═══════════════════════

    /** GET /days — list schedule days with optional filters */
    suspend fun getDays(startDate: Long? = null, endDate: Long? = null, typeId: String? = null): ApiResponse<List<ScheduleDay>> {
        return client.get("$baseUrl/days") {
            startDate?.let { parameter("startDate", it.toString()) }
            endDate?.let { parameter("endDate", it.toString()) }
            typeId?.let { parameter("typeId", it) }
        }.body()
    }

    /** POST /days — create a schedule day */
    suspend fun createDay(
        title: String, typeId: String, dateRangeType: String,
        calendarDays: List<Long>, timezone: String = "UTC",
        conflictAction: String = ""
    ): ApiResponse<ScheduleDay> {
        val response: HttpResponse = client.post("$baseUrl/days") {
            setBody(buildJsonObject {
                put("title", title)
                put("typeId", typeId)
                put("dateRangeType", dateRangeType)
                put("calendarDays", JsonArray(calendarDays.map { JsonPrimitive(it) }))
                put("startDate", calendarDays.minOrNull() ?: 0L)
                put("endDate", calendarDays.maxOrNull() ?: 0L)
                put("numberOfDays", calendarDays.size)
                put("timezone", timezone)
                put("conflictAction", conflictAction)
            })
        }
        // 409 Conflict: body is {status:0, message:"schedule_day_conflict", data:{conflicts:[...]}}
        if (response.status.value == 409) {
            val raw = Json.parseToJsonElement(response.bodyAsText()).jsonObject
            val conflicts = raw["data"]?.jsonObject?.get("conflicts")?.jsonArray
                ?.mapNotNull { it as? JsonObject }
                ?: emptyList()
            throw ScheduleConflictException(conflicts)
        }
        return response.body()
    }

    /** PUT /days/:id — update a schedule day */
    suspend fun updateDay(id: String, data: JsonObject): ApiResponse<ScheduleDay> {
        val response: HttpResponse = client.put("$baseUrl/days/$id") {
            setBody(buildJsonObject {
                data.forEach { key, value -> put(key, value) }
            })
        }
        if (response.status.value == 409) {
            val raw = Json.parseToJsonElement(response.bodyAsText()).jsonObject
            val conflicts = raw["data"]?.jsonObject?.get("conflicts")?.jsonArray
                ?.mapNotNull { it as? JsonObject }
                ?: emptyList()
            throw ScheduleConflictException(conflicts)
        }
        return response.body()
    }

    /** Atomic single-day type change — PUT /days/:id/single-date. */
    suspend fun updateSingleDay(id: String, date: Long, typeId: String, action: String) {
        client.put("$baseUrl/days/$id/single-date") {
            setBody(buildJsonObject {
                put("date", date)
                put("typeId", typeId)
                put("action", action)
            })
        }.body<ApiResponse<JsonObject>>()
    }

    /** DELETE /days/:id — no body; userId comes from moduledata. */
    suspend fun deleteDay(id: String): ApiResponse<JsonObject> {
        return client.delete("$baseUrl/days/$id").body()
    }

    /** POST /days/bulk — bulk update multiple days */
    suspend fun bulkUpdateDays(updates: List<JsonObject>): ApiResponse<List<ScheduleDay>> {
        return client.post("$baseUrl/days/bulk") {
            setBody(buildJsonObject {
                put("updates", JsonArray(updates))
            })
        }.body()
    }

    /** POST /days/remove-dates — remove specific dates from day blocks */
    suspend fun removeDates(entries: List<JsonObject>): ApiResponse<JsonObject> {
        return client.post("$baseUrl/days/remove-dates") {
            setBody(buildJsonObject {
                put("entries", JsonArray(entries))
            })
        }.body()
    }

    /** POST /days/duplicate */
    suspend fun duplicateDay(sourceDayId: String, newStartDate: Long): ApiResponse<ScheduleDay> {
        return client.post("$baseUrl/days/duplicate") {
            setBody(buildJsonObject {
                put("sourceDayId", sourceDayId)
                put("newStartDate", newStartDate)
            })
        }.body()
    }

    // ═══════════════════════ EVENTS ═══════════════════════

    /** GET /events */
    suspend fun getEvents(
        startDate: Long? = null, endDate: Long? = null,
        scheduleDayId: String? = null, eventType: String? = null
    ): ApiResponse<List<ScheduleEvent>> {
        return client.get("$baseUrl/events") {
            startDate?.let { parameter("startDate", it.toString()) }
            endDate?.let { parameter("endDate", it.toString()) }
            scheduleDayId?.let { parameter("scheduleDayId", it) }
            eventType?.let { parameter("eventType", it) }
        }.body()
    }

    /** POST /events — create event or note */
    suspend fun createEvent(data: JsonObject): ApiResponse<ScheduleEvent> {
        return client.post("$baseUrl/events") {
            setBody(buildJsonObject {
                data.forEach { key, value -> put(key, value) }
            })
        }.body()
    }

    /** PUT /events/:id */
    suspend fun updateEvent(id: String, data: JsonObject): ApiResponse<ScheduleEvent> {
        return client.put("$baseUrl/events/$id") {
            setBody(buildJsonObject {
                data.forEach { key, value -> put(key, value) }
            })
        }.body()
    }

    /** DELETE /events/:id — no body; userId comes from moduledata. */
    suspend fun deleteEvent(id: String): ApiResponse<JsonObject> {
        return client.delete("$baseUrl/events/$id").body()
    }

    // ═══════════════════════ CALENDAR ═══════════════════════

    /** GET /calendar — days with nested events/notes */
    suspend fun getCalendar(startDate: Long? = null, endDate: Long? = null): ApiResponse<List<ScheduleDay>> {
        return client.get("$baseUrl/calendar") {
            startDate?.let { parameter("startDate", it.toString()) }
            endDate?.let { parameter("endDate", it.toString()) }
        }.body()
    }

    // ═══════════════════════ ACTIVITY LOG ═══════════════════════

    /** GET /activity-log */
    @Serializable
    data class ActivityLogResponse(
        val logs: List<ActivityLog>,
        val total: Int,
        val page: Int,
        val limit: Int
    )

    suspend fun getActivityLog(
        limit: Int = 50, page: Int = 0,
        startDate: Long? = null, endDate: Long? = null
    ): ApiResponse<ActivityLogResponse> {
        return client.get("$baseUrl/activity-log") {
            parameter("limit", limit.toString())
            parameter("page", page.toString())
            startDate?.let { parameter("startDate", it.toString()) }
            endDate?.let { parameter("endDate", it.toString()) }
        }.body()
    }

    // ═══════════════════════ REVISIONS ═══════════════════════

    /** GET /revisions */
    suspend fun getRevisions(startDate: Long? = null, endDate: Long? = null): ApiResponse<List<Revision>> {
        return client.get("$baseUrl/revisions") {
            startDate?.let { parameter("startDate", it.toString()) }
            endDate?.let { parameter("endDate", it.toString()) }
        }.body()
    }

    /** GET /revisions/current */
    suspend fun getCurrentRevision(): ApiResponse<Revision> {
        return client.get("$baseUrl/revisions/current").body()
    }

    // ═══════════════════════ SHARE ═══════════════════════

    @Serializable
    data class ShareLinkResponse(val token: String, val shareUrl: String)

    /** POST /share/generate-link */
    suspend fun generateShareLink(): ApiResponse<ShareLinkResponse> {
        return client.post("$baseUrl/share/generate-link").body()
    }

    @Serializable
    data class SharedScheduleResponse(
        val projectId: String,
        val days: List<ScheduleDay>,
        val events: List<ScheduleEvent>,
        val types: List<ScheduleType>
    )

    /** GET /share/:token — public, no auth needed */
    suspend fun getSharedSchedule(token: String): ApiResponse<SharedScheduleResponse> {
        return client.get("$baseUrl/share/$token").body()
    }
}
